lib/android/app/src/main/java/com/reactnativenavigation/views/stack/topbar/titlebar/MainToolBar.kt

```diff
@@ -160,6 +142,69 @@ class MainToolBar(context: Context) : RelativeLayout(context) {
         clearComponent()
     }
 
+    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
+        super.onLayout(changed, l, t, r, b)
+        // Bypass regular layout for the titleComponent to make final placements based on the measurement
+        val titleComponent = getTitleComponent()
+        if (titleComponentAlignment == Alignment.Center) {
+            val titleLeft = max(l + leftButtonsBar.measuredWidth, (r - l) / 2 - titleComponent.measuredWidth / 2)
+            val titleRight = min(r - rightButtonsBar.measuredWidth, (r - l) / 2 + titleComponent.measuredWidth / 2)
+            titleComponent.layout(titleLeft, t, titleRight, b)
+        } else {
+            titleComponent.layout(l + leftButtonsBar.measuredWidth + defaultMargin, t, r - rightButtonsBar.measuredWidth, b)
+        }
+    }
+
+    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+        val mode = MeasureSpec.getMode(widthMeasureSpec)
+        val parentWidth = MeasureSpec.getSize(widthMeasureSpec)
+        val parentHeight = MeasureSpec.getSize(heightMeasureSpec)
+        logd("OnMeasure: parentWidth: $parentWidth, parentHeight: $parentHeight widthMeasureSpec:" +
+                " ${MeasureSpec.toString(widthMeasureSpec)}, heightMeasureSpec: ${MeasureSpec.toString(heightMeasureSpec)}", "MainToolBar")
+        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
+        // exact measurement of the parent
+        if (mode == MeasureSpec.EXACTLY) {
+            val titleComponent = getTitleComponent()
+            // measure all 3 child with AT_MOST to get what space they require
+            rightButtonsBar.measure(MeasureSpec.makeMeasureSpec(parentWidth, MeasureSpec.AT_MOST), heightMeasureSpec)
+            leftButtonsBar.measure(MeasureSpec.makeMeasureSpec(parentWidth, MeasureSpec.AT_MOST), heightMeasureSpec)
+            titleComponent.measure(MeasureSpec.makeMeasureSpec(parentWidth, MeasureSpec.AT_MOST), heightMeasureSpec)
+
+            val titleComponentWidth = titleComponent.measuredWidth
+
+            // hide buttons title bypass the right end, otherwise show
+            var leftBarWidth = leftButtonsBar.measuredWidth
+            val titleLeft = parentWidth / 2 - (titleComponentWidth / 2)
+            if (titleLeft < this.left + leftBarWidth + defaultMargin) {
+                leftBarWidth = minToolBarSize
+                for (item in leftButtonsBar.menu) {
+                    item.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
+                }
+            }else{
+                for (item in leftButtonsBar.menu) {
+                    item.setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
+                }
+            }
+
+            // hide buttons title bypass the left end, otherwise show
```

Comment:
Duplicate code - duplicate comments 😅

Reply:
I left the design as-is for now, to think about the design and enhancing the code, I'm bad at naming stuff so I left them laid like this in the code, while I'm having my time designing it to look right :]